File: csharp/ToolGood.Algorithm/Internals/Functions/Operator/Function_Sub.cs
using System;
using System.Text;

namespace ToolGood.Algorithm.Internals.Functions.Operator
{
	internal sealed class Function_Sub : Function_2
	{
		public Function_Sub(FunctionBase[] funcs) : base(funcs)
		{
		}

		public Function_Sub(FunctionBase func1, FunctionBase func2) : base(func1, func2)
		{
		}

		public override string Name => "-";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args1 = GetNumber_1(engine, tempParameter); if(args1.IsError) { return args1; }
			var args2 = GetNumber_2(engine, tempParameter); if(args2.IsError) { return args2; }

			if(args2.NumberValue == 0m) { return args1; }

			return Operand.Create(args1.NumberValue - args2.NumberValue);
		}
		public override void ToString(StringBuilder stringBuilder, bool addBrackets)
		{
			if(addBrackets) stringBuilder.Append('(');
			func1.ToString(stringBuilder, false);
			stringBuilder.Append(" - ");
			func2.ToString(stringBuilder, false);
			if(addBrackets) stringBuilder.Append(')');
		}
	}
}

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/operator/Function_Sub.java
package toolgood.algorithm.internals.functions.operator;

import java.math.BigDecimal;
import java.util.List;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_2;

public final class Function_Sub extends Function_2 {

    public Function_Sub(FunctionBase[] funcs) {
        super(funcs);
    }

    public Function_Sub(FunctionBase func1, FunctionBase func2) {
        super(func1, func2);
    }

    @Override
    public String Name() {
        return "-";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetNumber_1(engine, tempParameter); if (args1.IsErrorOrNone()) { return args1; }
        Operand args2 = GetNumber_2(engine, tempParameter); if (args2.IsErrorOrNone()) { return args2; }

        if (args2.NumberValue().compareTo(BigDecimal.ZERO) == 0) { return args1; }

        return Operand.Create(args1.NumberValue().subtract(args2.NumberValue()));
    }

    @Override
    public void ToString(StringBuilder stringBuilder, boolean addBrackets) {
        if (addBrackets) stringBuilder.append('(');
        func1.ToString(stringBuilder, false);
        stringBuilder.append(" - ");
        func2.ToString(stringBuilder, false);
        if (addBrackets) stringBuilder.append(')');
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}
